src/library/searchquery.cpp

```diff
@@ -166,6 +173,36 @@ QString TextFilterNode::toSql() const {
     return concatSqlClauses(searchClauses, "OR");
 }
 
+CrateFilterNode::CrateFilterNode(const QSqlDatabase& database,
+                                 const QString& argument,
+                                 const CrateStorage* pCrateStorage)
+    : m_database(database),
+      m_argument(argument),
+      m_pCrateStorage(pCrateStorage),
+      m_matchInitialized(false) {
+}
+
+bool CrateFilterNode::match(const TrackPointer& pTrack) const {
+    if (!m_matchInitialized) {
+        CrateTrackSelectResult crateTracks(
+             m_pCrateStorage->selectTracksSortedByCrateNameLike(m_argument));
+
+        while (crateTracks.next()) {
+            m_matchingTrackIds.push_back(crateTracks.trackId());
+        }
+
+        m_matchInitialized = true;
+    }
+
+    return std::binary_search(m_matchingTrackIds.begin(), m_matchingTrackIds.end(), pTrack->getId());
+}
+
+QString CrateFilterNode::toSql() const {
+    return QString("id IN (%1)").arg(CrateStorage::formatQueryForTrackIdsByCrateNameLike(
```

Comment:
If you make formatQueryForTrackIdsByCrateNameLike() a non-static member function of CrateStorage you can get rid of the awkward database parameter and member of CrateFilterNode. You are storing a pointer to CrateStorage anyway which already carries a database connection.